File: src/core/Crypto/Token.cs
using System;

namespace core.Crypto
{
    public struct Token
    {
        private readonly string _symbol;

        public Token(string symbol)
        {
            if (string.IsNullOrWhiteSpace(symbol))
            {
                throw new ArgumentException(nameof(symbol), "Symbol cannot be blank");
            }
            _symbol = symbol.ToUpper();
        }

        public static implicit operator string(Token t) => t._symbol;
        public static implicit operator Token(string t) => new Token(t);

        public string Value => _symbol;
    }
}

File: src/core2.fs/Cryptos/Token.fs
namespace core.Cryptos

open System

[<Struct>]
type Token =
    val private _symbol: string
    
    new (symbol: string) =
        if String.IsNullOrWhiteSpace(symbol) then
            raise (ArgumentException("Symbol cannot be blank", "symbol"))
        { _symbol = symbol.ToUpper() }

    static member op_Implicit(t: Token) : string = t._symbol
    static member op_Implicit(t: string) : Token = Token(t)

    member this.Value = this._symbol
